### ISSUE TITLE
Script for grounding/normalising targets and diseases extracted from literature using OTAR APIs

### ISSUE THREAD

**Issue by tsantosh7** [opentargets/platform#996]
Please generate a script that takes in a JSON series of targets (genes/proteins) and diseases and provides a database ID reference (grounding/normalisation) utilising the OTAR APIs. 

The input for the script can be found at https://drive.google.com/file/d/1jx9hNCIE7hsnJAlmq3W-xEvolv2taggZ/view?usp=sharing

discarded format: https://drive.google.com/open?id=1NdVamYvHg32C6tvuaHTiFsSXas0Qh4vO

**Comment by d0choa** [comment]
Hi @tsantosh7. I hope you are well. We might start looking at this task this week. 

I'm assuming the above file is for cases where GP and DS co-occur. We talked about generating this file before the target-disease co-occurrence. In that case, we could have papers where only 1 of the entities occur. Also it would be good to be able to include the "drug/chemical" entity in a sample file.

Could you generate a similar sample file to the above with these characteristics? We will in the meantime use the previous file.

**Comment by tsantosh7** [comment]
Hi @d0choa  Thank you and hope you are doing well too! 

Yes the above file is where GP and DS co-occur. Yes, then It will then be after the ML False positive removal. I will try to include drug/chemical and generate a new jsonl file with the aforementioned characteristics. Meanwhile please try using the previous file to get things started.

**Comment by mkarmona** [comment]
@tsantosh7 should I expect any string normalisation before generating this file? I got a random example and I can see entities are not clean enough

```
{
  "PMC1241207": {
    "GP": [
      "T-helper type 1",
      "T(H)1",
      "NF-Kappa B",
      "NF-Kappa B.",
      "nuclear factor-kappaB"
    ],
    "DS": [
      "infections",
      "viral upper respiratory tract infections",
      "nonallergic",
      "Asthma",
      "asthma"
    ]
  }
}
```

**Comment by tsantosh7** [comment]
@mkarmona there is no string normalisation in place. Those strings are extracted as how they appear in the text written by the authors. Do you want me to group similar strings together if it makes it easy for you?

@d0choa Between I am working on extracting drugs/chemicals as well. So @mkarmona  will receive another sample file with the addition of drugs/chemicals by end of this week.

**Comment by mkarmona** [comment]
@tsantosh7 you can forget about doing any string transformation but the JSON you are outputting is not the most suitable for me. Can you output in this way? please not that you are producing a dump of a dict where the key is the value itself and I need to have a proper common key, by example, `pmid` so now the serialised json (that is pretty-printed, I don't need it pretty-printed is just to make it more readable but rather serialised as you are currently doing) is an object each line where the keys are `pmid`, `GP` and `DS` and perhaps another for the list of drugs tokens. Thanks
```
{
  "pmid": "PMC1241207",
  "GP": [
      "T-helper type 1",
      "T(H)1",
      "NF-Kappa B",
      "NF-Kappa B.",
      "nuclear factor-kappaB"
    ],
  "DS": [
      "infections",
      "viral upper respiratory tract infections",
      "nonallergic",
      "Asthma",
      "asthma"
  ]
}
```

**Comment by mkarmona** [comment]
@tsantosh7 it could be great if you generate the same file but in the way I need it to keep working on it while you finish the one with drugs included. thanks in advance!

**Comment by tsantosh7** [comment]
@mkarmona the updated file in the format you have asked for: https://drive.google.com/file/d/1NdVamYvHg32C6tvuaHTiFsSXas0Qh4vO/view?usp=sharing

**Comment by mkarmona** [comment]
@tsantosh7 here you have a run from the last file you gave me
[mapped_terms.jsonl.gz](https://github.com/opentargets/platform/files/5300107/mapped_terms.jsonl.gz)

**Comment by tsantosh7** [comment]
:+1:  Thank you @mkarmona. I will have a discussion with @aravindvenkatesan and get back to you

**Comment by mkarmona** [comment]
@tsantosh7 I generated another file. this is another iteration that contains drug filtering. few things I got. thanks!

- you are treating cells and drugs as targets

[mapped_terms.jsonl.gz](https://github.com/opentargets/platform/files/5300633/mapped_terms.jsonl.gz)

**Comment by mkarmona** [comment]
@tsantosh7 I improved the mapping and I am also including mapped terms separated with raw data as you can see where is that coming from. I do not interfere in your detection so I separated direct mappings as entities you classify from the ones I detect crossed entity

[mapped_terms.jsonl.gz](https://github.com/opentargets/platform/files/5304606/mapped_terms.jsonl.gz)

**Comment by tsantosh7** [comment]
Thank you @mkarmona for giving us 3 options. I have a meeting with @aravindvenkatesan in few hours from now. I will discuss with him and come up with a format. For now I think the first option looks easy to map the terms from our side. Anyways I will update you before the close of play today.

**Comment by mkarmona** [comment]
Perfect @tsantosh7 . Please consider this latest file as we think it could be the most appropriate for all. IMHO i think is the less opinionated of the all attempts

**Comment by tsantosh7** [comment]
@mkarmona I just compared all the three. I agree   :+1:  that the last option looks good too. However I will confirm later today.

**Comment by mkarmona** [comment]
@tsantosh7 here a bit more elaborated file and what each field means
[mapped_terms.jsonl.gz](https://github.com/opentargets/platform/files/5306486/mapped_terms.jsonl.gz)

```
{
  "pmid": "PMC1241358",
  "terms_mapped": [
    {
      "term_raw": "allergy",
      "term_norm": "allergy",
      "id": "EFO_0003785",
      "term_type": "disease",
      "keyword_type": "disease"
    }
  ],
  "terms_not_mapped": [
    {
      "term_raw": "IgE antibody",
      "term_norm": "igeantibody",
      "term_type": "target"
    },
...
    {
      "term_raw": "allergic diseases",
      "term_norm": "allergicdiseases",
      "term_type": "disease"
    }
  ],
  "targets_mapped": [],
  "diseases_mapped": [
    {
      "term_raw": "allergy",
      "term_norm": "allergy",
      "id": "EFO_0003785",
      "term_type": "disease",
      "keyword_type": "disease"
    }
  ],
  "drugs_mapped": [],
  "cross_mapped": []
}

```

- `pmid` pubmed id unique field across the dataset
- `terms_mapped` the full list of terms provided and the full list of mapped entities to dictionaries whether their types are the same
- `terms_not_mapped` list of terms were not possible to map to any element of the full dictionary
- `targets_mapped` subset of `terms_mapped` when both types are the same and `== target`
- `diseases_mapped` subset of `terms_mapped` when both types are the same and `== disease`
- `drugs_mapped` subset of `terms_mapped` when both types are the same and `== drug` (currently not used)
- `cross_mapped` the ones mapped but their types are not the same

`term_raw` and `term_type` (already classified in the input file) are provided by the input file

**Comment by tsantosh7** [comment]
Hi @mkarmona and @d0choa 

Thank you very much, this is working so well. From our analysis today, here is our feedback:

Example considered: PM_ID: PMC1240767

Option_1: We cannot compare original terms with the normalised terms as they are different. E;g: APLP-2 vs aplp2 and Alzheimer disease vs alzheimerdisease

Option_2: Its nice to see the drug filter added to the json stream. However it has the demerits listed from option-1, i.e., missing of the raw terms.

Option_3: This format solves the aforementioned issues. The raw terms (original terms) are mapped to the normalised ones, making it easier to compare and link at end our end. However, it would be nice if you could add all the list of entities that were originally recieved by the API. This will enable us to do some validations later on.

Other issues:
1. What aboutthose terms that do not have entity linking? for E.g.: neuronal degeneration ?  Should we store all those missed entities into a database?

So, to summarise, as @mkarmona suggested option_3 works well for both your team and our team. We would request you to add the originally received entities as well just to make the validations easier when post processing.

**Comment by mkarmona** [comment]
Great @tsantosh7 . Thanks. You can find in the latest iteration I commented I solved the issue you mention so your initial provided terms are there.

**Comment by tsantosh7** [comment]
Thank you @mkarmona. Perfect!!

**Comment by tsantosh7** [comment]
Hi @mkarmona 

cced - @aravindvenkatesan

Thank you for the amazing stuff. I have few production level questions: 

1. How would I communicate with your service? 
2. Would you have a REST endpoint, where I post my jsonl.file? if so, the output of your end point could be a jsonified dict of json lines? 

Thank you!

**Comment by mkarmona** [comment]
Hi, @tsantosh7 and @aravindvenkatesan thanks! firstly,

It is currently an Apache Spark script that runs anywhere (the requirements are just a really few). It is [currently alpha status](https://github.com/opentargets/platform-etl-backend/blob/mk-assocs-graph/scripts/opentargets-epmc-entity-resolution.sc) so 

1. it is not properly wrapped in a standalone jar that anyone can run anywhere whether you run locally or run in a Hadoop cluster supporting Apache Spark 3.x without any modification, and 
2. it will be placed in its own repo where you can easily download the produced jar.

The way it works it takes 2 inputs where the lookup tables are and where the input files to transform (this is the one you generates @tsantosh7 ) and specify an output folder too. The output format is a dataset composed of multiple files in a folder where each file contains serialised JSON-lines. that way you can further consume this output for post-processing at scale.

The output you are generating to feed this script 

1. has to be serialised JSON-lines but not the way you gave it initially. it needs to use common keys. 
2. it does not matter you produce one or many files but ensuring you meet with the serialised JSON-lines

This [comment](https://github.com/opentargets/platform/issues/996#issuecomment-700692621) clarifies this point further

Regarding your question about the place to put the dataset you produce it could be temporally any that can be accessed in order to download and ensure it does all properly. you could potentially upload to some of our GCS buckets

**Comment by tsantosh7** [comment]
@mkarmona Thank you, please give me a weeks time up with some practicality issues.

**Comment by mkarmona** [comment]
@tsantosh7 sure! Please tell me if I can help with any scalability or etl related issue. Here you have a pair of hands if needed

**Comment by tsantosh7** [comment]
@mkarmona  and @d0choa Sorry for getting back late. I am still working on extracting the drugs and chemicals from all the data. Now that I have datasets created for it, will try to develop model today. On the other hand, I had been thinking about the integration of @mkarmona normalisation program into this architecture. Previously I thought its a rest API service but then it seems  different to what I have planned. Nevertheless, I feel this is good in the sense that its faster than calling it as a service. In addition to  this, I am more concerned about the accuracy as well. For instance, the normalised terms do not care for any punctuation.  I would like to know more about the normalisation process please.

if you are available @mkarmona could we have a brief meeting on Tuesday? I am available anytime from 9 to 5:30.

**Comment by mkarmona** [comment]
@tsantosh7 yes happy to meet on Tuesday. please place gmeet on the gcalendar and count me in. Regarding 

> I am more concerned about the accuracy as well. For instance, the normalised terms do not care for any punctuation. I would like to know more about the normalisation process

Can you please include some example about this to discuss about? I mean, the transformation I do are quite minimal and try to keep the nature of the word but things like upercase, spaces, ending punctuations or excesive parentheses stop the pure string match working. I am not currently implementing any fuzzy / similarity algorithm to match as I dont want to put another variable into the system.

The code that clean the sentence is quite simple though
```Scala
def normalise(c: Column): Column = 
  lower(translate(trim(trim(c), "."), "[]{}()'- ", ""))
```
@tsantosh7 do you think that makes sense remove plurals in order to match with a disease?

**Comment by mkarmona** [comment]
@tsantosh7 it could great if you can generate a full set without drugs to check performance and acurracy on my side at least. The more iterations we do the better the integration and the sooner we catch potential problems.

**Comment by tsantosh7** [comment]
@mkarmona Thanks for this. I will generate without drugs first. You will have this ready by Monday.

**Comment by tsantosh7** [comment]
> @tsantosh7 yes happy to meet on Tuesday. please place gmeet on the gcalendar and count me in. Regarding
> 
> > I am more concerned about the accuracy as well. For instance, the normalised terms do not care for any punctuation. I would like to know more about the normalisation process
> 
> Can you please include some example about this to discuss about? I mean, the transformation I do are quite minimal and try to keep the nature of the word but things like upercase, spaces, ending punctuations or excesive parentheses stop the pure string match working. I am not currently implementing any fuzzy / similarity algorithm to match as I dont want to put another variable into the system.
> 
> The code that clean the sentence is quite simple though
> 
> ```scala
> def normalise(c: Column): Column = 
>   lower(translate(trim(trim(c), "."), "[]{}()'- ", ""))
> ```
> 
> @tsantosh7 do you think that makes sense remove plurals in order to match with a disease?

**Comment by tsantosh7** [comment]
@mkarmona Please find the GP and DS tags from all the articles: https://drive.google.com/file/d/1IpOK5hj1DnWdr3ycrs_ztnOOOgAqGtWN/view?usp=sharing

**Comment by mkarmona** [comment]
thanks, @tsantosh7 ! I already mapped them. It took a couple of minutes in my laptop but it generates 14GB in total so I am compressing them and uploading to drive to share the results with you. Unfortunately, It will take more time to upload than compute it.

**Comment by tsantosh7** [comment]
That was pretty quick @mkarmona. I will try to send you drugs as early as possible.

**Comment by mkarmona** [comment]
@tsantosh7 here the first [mapping pass](https://storage.googleapis.com/ot-team/carmona/terms_mapped_20201013.tar)

**Comment by tsantosh7** [comment]
Thank you @mkarmona

**Comment by mkarmona** [comment]
@aravindvenkatesan @tsantosh7 [here some counts](https://docs.google.com/spreadsheets/d/1KJpBHpnxzlhzYEG8Z6B3QVm1qOHz8p13pmEUSIzT8yA/edit?usp=sharing) from the resulted mappings

**Comment by d0choa** [comment]
A problem with `intrinsic` being picked up by ML and grounded by this pipeline, due to the word `intrinsic` showing as a synonym of the disease has been reported to [MONDO](https://github.com/monarch-initiative/mondo/issues/2122)

**Comment by tsantosh7** [comment]
@d0choa There was no ML filtering for disease. So, it was purely the dictionary output. ML tagged it as a whole "urethral intrinsic sphincter deficiency". We could discuss about this in our monthly meeting.

Between, @mkarmona's normalisation is working well. This will open our analysis to introducing ML based annotations as well into pipeline, which are then normalised using @mkarmona script. Very exciting stuff!!

**Comment by mkarmona** [comment]
@tsantosh7 thanks. The moment I have drugs incorporated I will also commit few improvements as plurals and Greek letters transformations. These two fixes will bring few new terms detected many times across literature.

**Comment by tsantosh7** [comment]
@mkarmona cheers, I am on it.

**Comment by tsantosh7** [comment]
@mkarmona and @d0choa, as discussed, Here are the abstracts with GP, DS and CD tags https://drive.google.com/file/d/1OyvzMk971iX9d5Ohx_tSCEMgBqeUyWPb/view?usp=sharing

Something like this:

```
{
  "pmid": "PMID31911178",
  "GP": [
    "NR2B",
    "Bcl-2",
    "Bax",
    "p38"
  ],
  "DS": [
    "MMPs",
    "depression"
  ],
  "CD": [
    "ferulate",
    "glutamate",
    "Coniferyl ferulate",
    "lactate",
    "malondialdehyde"
  ]
}
```

Another example

```
{
  "pmid": "PMID31911267",
  "GP": [
    "TNF-α",
    "IL-6",
    "IL-1β"
  ],
  "DS": [
    "neurodegenerative diseases"
  ],
  "CD": [
    "princepin",
    "americanin B",
    "sesquineolignans",
    "diverniciasin B",
    "dineolignans",
    "isoprincepin",
    "isoamericanin A",
    "phenylpropanoids",
    "neolignans",
    "diverniciasin C",
    "isodiverniciasin A",
    "monolignan",
    "steroid"
  ]
}


```
The model was not scaling to Fulltext articles due to the size issue. I just understood on how to solve this bug. So I will fix and try to send you Fulltext tags, close to end of play today.

Known Limitations with CD tags:
Due to a tokenizing issue while training, if there is a special character between a term, sometimes, it treats as two terms. It is happening with '('  and '-'. I will need to fix this but will do it next week and regenerate these tags again.

@mkarmona  let me know how you find the CDs. 

Best wishes
Santosh

**Comment by mkarmona** [comment]
@tsantosh7 [here](https://storage.googleapis.com/ot-team/carmona/mapped_entities_20201016.tar) the results. it looks quite good!

**Comment by mkarmona** [comment]
@tsantosh7 nice improvement!
```scala
@ terms.filter(size($"targets_mapped") > 0 and size($"diseases_mapped") > 0 and size($"drugs_mapped") > 0).count 
res22: Long = 1233637L
@ terms.filter(size($"targets_mapped") === 0 and size($"diseases_mapped") > 0 and size($"drugs_mapped") > 0).count 
res23: Long = 235435L
@ terms.filter(size($"targets_mapped") > 0 and size($"diseases_mapped") === 0 and size($"drugs_mapped") > 0).count 
res24: Long = 585695L
@ terms.filter(size($"targets_mapped") > 0 and size($"diseases_mapped") > 0).count 
res25: Long = 2996961L
@ terms.filter(size($"targets_mapped") > 0 and size($"diseases_mapped") > 0 and size($"terms_not_mapped") === 0).count 
res26: Long = 1812479L
@ terms.filter(size($"targets_mapped") > 0 and size($"diseases_mapped") > 0 and size($"terms_not_mapped") > 0).count 
res27: Long = 1184482L
```

**Comment by tsantosh7** [comment]
@mkarmona Just to update you that the program is still running on the Fulltext articles. Its been 66hrs and only 60% of the articles have been processed. I will try to give you by tomorrow hoping they get processed. 

@d0choa I learnt some lessons during this, I can talk about them during our meeting this month.

1. - 1. Running the CD model (or GP, DS) on each sentence is time-consuming on the Fulltext. So it is best to run the DL model on only those that have GP or DS tags. 
2. So, if we would like to introduce FNs from GP, DS,  it is better to combine the GP, DS and CD datasets and train a new lighter model. In the next version, we can think about this. 
3. For this version, it is better we run the CD model on only those sentences that have GP or DS tags.

**Comment by mkarmona** [comment]
Please @tsantosh7 can you reply with a simple list of your terms and what those stand for?

> CD, GP, DS, FN....

**Comment by tsantosh7** [comment]
@mkarmona Sure

CD: Chemical/Drug
GP: Gene/Protein
DS: Disease
FN: False negatives (The ones that are missed by dictionaries)

**Comment by AsierGonzalez** [comment]
What about `DL`?

**Comment by d0choa** [comment]
We can iterate, but we will eventually need to run the CD model on all sentences (not only GP or DS tagged sentences).
It's the only way we can list the articles that mention a particular drug/chemical isolated or in the context of other entities

**Comment by tsantosh7** [comment]
@AsierGonzalez DL is Deep learning. I now have replaced it with CD to make it specific to the context here.

**Comment by tsantosh7** [comment]
@d0choa Thank you. I totally understand. It takes a lot of extra time but nonetheless it's doable. For our next iteration, it is best that we produce 1) one standard model by fusing CD dataset with GP and DS, and 2) a lighter version of the current DL model.

**Comment by tsantosh7** [comment]
@mkarmona it took more than 5 days to process 90% of the fulltext. Here is the file https://drive.google.com/file/d/10k4q1t0ZxOusZhe3fYmyQW7mTjOMp8-Y/view?usp=sharing

Kindly run your analysis on this while I will investigate ways to make the model run faster.


Here is the time taken for processing all the Fulltext

```
Files  Time
==== =====
7799 81:01:52
4838 58:16:31
7373 65:02:37
8526 45:08:55
6506 41:31:23
4125 30:39:53
4920 21:58:38
3621 11:38:53
4656 21:33:38
6834 19:30:18
6485 44:07:10
7303 42:44:03
7145 78:30:28
5730 82:02:34
6584 86:16:31
4982 66:11:57
16%|█▌ 
8496 67:11:56
8278 69:15:09
7933 69:08:58
8437 47:09:20
8662 133:21:10
8799 48:57:57
8656 100:56:25
8694 83:39:32
8615 80:55:09
8062 65:23:44
7644 73:19:03
5668 25:16:23
8256 83:16:30
8603 75:11:38
7964 80:38:07
8467 46:48:32
8517 58:04:24
93%|█████████▎| 8181/8838
8845 53:04:46
8712 79:54:13
8761 65:24:05
8789 92:53:00
8170 89:28:33
8505 44:15:16
8821 91:24:43
7621 41:04:24
6985 45:56:15
8284 54:24:57
8356 66:07:05
8103 65:36:04
8623 138:00:20
8377 64:03:08
8679 86:36:48
8593 44:42:08
8293 35:46:48
8228 76:07:56
8552 44:17:43
8345 61:48:26
8118 108:19:55
8157 58:41:51
8056 47:44:33
8701 104:49:28
8664 66:16:14
8629 50:12:02
8181 46:51:10
8326 89:55:32
8309 43:58:07
8385 69:14:27
8234 129:23:04
8710 38:34:53
8622 124:26:41
8601 48:09:39
8765 92:01:43
8264 122:20:49
8731 93:53:51
8526 112:27:20
8625 68:35:38
8581 63:16:47
8294 57:56:26
8658 68:33:16
96%|█████████▌| 8336/8697
8476 95:27:46
8149 75:01:19
93%|█████████▎| 7245/7776
8431 80:14:01
7858 44:28:10
7988 62:51:32
7380 86:27:41
7609 116:05:51
8266 89:46:13
8674 51:03:45
8786 112:42:29
7386 42:14:43
97%|█████████▋| 8317/8543
89%|████████▉ |
8395 61:04:06
8682 80:05:36
8529 87:31:54
8645 90:32:34
8687 45:43:06
8402 90:20:16
8668 137:53:26
66%|██████▋ 
7512 52:23:22
7759 114:21:24
8459 55:45:29
5994 39:39:36
8397 54:56:00
8492 90:14:47
6399 90:21:53
8208 100:49:44
8275 44:42:24
83%|████████▎ |
8271 83:33:48
8677 57:30:00
8381 115:25:42
8268 51:37:35
8557 71:05:13
8307 51:39:16
8591 42:41:36
8433 49:45:39
7733 75:25:35
8201 61:22:24
8442 74:51:13
8664 59:11:33
45%|████▍ 
86%|████████▌ |
95%|█████████▌| 7991/8379
5708 54:43:37
5485 30:13:30
8282 130:18:10
8415 68:22:29
7894 54:31:20
6401 65:08:40
8557 54:00:14
4158 36:20:05
7318 66:55:34
2106 21:37:56
8254 64:40:23
7866 111:43:10
8014 48:09:14
63%|██████▎ 
96%|█████████▌| 7161/7456
32%|███▏ 
98%|█████████▊| 7127/7280
45%|████▍ 
6356 91:42:08
2148 45:24:41
90%|█████████ |
5502 52:13:29
3404 35:18:53
3549 36:48:22
2711 22:34:29
2331 32:25:04
2617 18:33:58
2445 36:31:55
5443 38:29:06
2795 17:01:28
3957 43:19:04
2371 66:18:53
2934 28:07:07
2100 50:03:37
6668 80:29:51
5555 38:57:11
5977 61:42:51
68%|██████▊ 
7301 77:06:55
7794 124:08:52
6794 40:41:56
7683 47:05:58
3263 19:50:15
7594 49:05:02
5135 49:14:14
7635 47:36:23
8028 81:07:16
7849 69:15:37
4653 20:52:07
7143 45:31:11
7868 53:38:49
7679 73:46:52
7139 67:44:03
8399 52:46:12
6924 47:43:45
4078 43:41:30
5025 58:37:54
94%|█████████▎| 1555/1663
5901 54:18:43
7815 65:36:41
77%|███████▋ 
6911 121:48:13
7889 47:51:37
74%|███████▍ 
7930 118:31:25
6966 50:04:52
20%|█▉ 
7757 129:21:23
66%|██████▋ 
90%|█████████ |
7818 50:06:05
95%|█████████▌| 7546/7937
3765 34:47:56
38%|███▊ 
7627 60:50:15
93%|█████████▎| 6951/7480
7391 50:23:27
94%|█████████▎| 6731/7186
6356 61:32:55
6981 55:00:43
74%|███████▍ 
6682 48:56:40
6406 67:03:45
4216 28:36:45
82%|████████▏ |
30%|███ 
75%|███████▌ 
6965 44:48:29
7155 97:02:23
7726 125:45:15
8070 101:20:31
6625 41:43:15
7752 51:24:05
6674 43:22:58
8249 52:21:04
7289 102:00:57
6396 70:33:30
7150 51:19:20
62%|██████▏ 
36%|███▌ 
4979 100:23:47
8056 55:05:35
8024 66:13:33
6005 70:48:58
7036 77:00:16
7784 45:18:02
6766 40:57:45
7725 67:13:30
7427 77:22:11
6493 105:13:18
7213 51:37:57
6439 43:27:29
5713 60:11:25
1393 15:27:14
6220 46:10:37
7170 55:19:52
7630 116:23:07
6513 115:22:45
6985 108:39:13
61%|██████▏ 
3457 58:58:31
95%|█████████▍| 6667/7052
2069 35:14:53
3895 36:09:40
6887 137:13:15
6827 54:36:06
 0%|
67%|██████▋ 
6582 88:29:00
6458 59:34:46
6824 50:27:25
7193 52:55:53
6344 72:45:12
7910 64:59:26
7312 52:21:06
5854 37:06:11
6843 62:44:22
7304 56:28:52
14%|█▍ 
5570 41:55:53
6441 63:57:10
6109 83:57:09
7320 104:21:57
49%|████▉ 
6821 74:54:26
5502 49:24:15
8046 55:17:52
62 22:22
7025 55:52:39
6433 74:29:35
5842 93:10:38
6070 44:17:30
6666 66:59:21
53%|█████▎ 
6487 53:24:02
5968 56:26:19
7822 49:31:56
7642 53:52:42
6639 101:22:46
6188 36:49:29
6086 53:42:16
6276 66:12:59
55%|█████▍ 
6472 64:35:55
1169 15:46:16
8549 50:43:55
5725 35:46:23
6663 42:54:26

```

**Comment by mkarmona** [comment]
[here the result](https://storage.googleapis.com/ot-team/carmona/mapped_entities_fulltext_20201023.tar) of the full papers output and some conditional counts

```scala
@ terms.filter(size($"targets_mapped") > 0 and size($"diseases_mapped") > 0 and size($"drugs_mapped") > 0).count 
res3: Long = 949540L

@ terms.filter(size($"targets_mapped") === 0 and size($"diseases_mapped") > 0 and size($"drugs_mapped") > 0).count 
res4: Long = 318032L

@ terms.filter(size($"targets_mapped") > 0 and size($"diseases_mapped") === 0 and size($"drugs_mapped") > 0).count 
res5: Long = 24724L

@ terms.filter(size($"targets_mapped") > 0 and size($"diseases_mapped") > 0).count 
res6: Long = 1092721L

@ terms.filter(size($"targets_mapped") > 0 and size($"diseases_mapped") > 0 and size($"terms_not_mapped") === 0).count 
res7: Long = 94244L

@ terms.filter(size($"targets_mapped") > 0 and size($"diseases_mapped") > 0 and size($"terms_not_mapped") > 0).count 
res8: Long = 998477L

@ terms.count 
res9: Long = 1929399L
```

**Comment by tsantosh7** [comment]
@mkarmona would you have time next week (Monday) for a call with me and Aravind?

**Comment by ktsirigos** [comment]
@d0choa is there still work pending on this one?
